File: windows/Passknight/Services/Generator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Passknight.Services
{
    /// <summary>
    /// Generate secure random passwords.
    /// <a href="https://github.com/bitwarden/clients/blob/main/libs/common/src/tools/generator/password/password-generation.service.ts"/>
    /// </summary>
    class Generator
    {
        private string lowercase = "abcdefghijklmnopqrstuvwxyz";
        private string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        private string numbers = "0123456789";
        private string symbols = "!@#$%^&*";

        private Random random = new Random();

        private string ShufflePositions(string positions)
        {
            var shuffled = positions.ToCharArray();
            for(int i = 0; i < shuffled.Length; i++)
            {
                int j = random.Next(0, shuffled.Length - 1);
                char tmp = shuffled[i];
                shuffled[i] = shuffled[j];
                shuffled[j] = tmp;
            }

            return new string(shuffled);
        }

        public string GeneratePassword(GeneratorSettings settings)
        {
            string positions = "";

            string characters = string.Empty;
            if (settings.Lowercase)
            {
                characters += lowercase;
                positions += 'l';
            }
            if(settings.Uppercase)
            {
                characters += uppercase;
                positions += 'u';
            }
            if (settings.Numbers)
            {
                characters += numbers;
                positions += 'n';
            }
            if (settings.Symbols)
            {
                characters += symbols;
                positions += 's';
            }

            while (positions.Length < settings.Length)
                positions += 'a';

            var shuffledPositions = ShufflePositions(positions);

            string password = "";
            for(int i = 0; i < settings.Length; i++)
            {
                string chars = characters;
                switch(shuffledPositions[i])
                {
                    case 'l':
                        chars = lowercase; break;
                    case 'u':
                        chars = uppercase; break;
                    case 'n':
                        chars = numbers; break;
                    case 's':
                        chars = symbols; break;
                }

                int rand = random.Next(0, chars.Length - 1);
                password += chars[rand];
            }

            return password;
        }
    }
}

File: android/app/src/main/java/com/example/passknight/models/Generator.kt
package com.example.passknight.models

import android.util.Log
import androidx.lifecycle.MutableLiveData
import kotlin.random.Random

/**
 * Generate random passwords using custom settings.
 * https://github.com/bitwarden/clients/blob/main/libs/common/src/tools/generator/password/password-generation.service.ts
 */
class Generator() {

    var lowercase: Boolean = true
        set(value) {
            field = value
            updatePassword()
        }

    var uppercase: Boolean = true
        set(value) {
            field = value
            updatePassword()
        }

    var numbers: Boolean = true
        set(value) {
            field = value
            updatePassword()
        }

    var symbols: Boolean = true
        set(value) {
            field = value
            updatePassword()
        }

    var length: Float = 15f
        set(value) {
            field = value
            updatePassword()
        }

    val generatedPassword: MutableLiveData<String> = MutableLiveData("")

    object CharSets {
        const val lowercase = "abcdefghijklmnopqrstuvwxyz"
        const val uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        const val numbers = "0123456789"
        const val symbols = "!@#\$%^&*"
    }

    init { updatePassword() }

    /**
     * @return A generated password using the current settings
     */
    fun generatePassword(): String {
        var positions = ""
        var characters = ""

        if(!lowercase && !uppercase && !symbols && !numbers) {
            return ""
        }

        if(lowercase) {
            characters += CharSets.lowercase
            positions += 'l'
        }
        if(uppercase) {
            characters += CharSets.uppercase
            positions += 'u'
        }
        if(numbers) {
            characters += CharSets.numbers
            positions += 'n'
        }
        if(symbols) {
            characters += CharSets.symbols
            positions += 's'
        }

        positions += "a".repeat((length - positions.length).toInt())

        val shuffled = positions.toCharArray()
        shuffled.shuffle()

        var password = ""
        for(char: Char in shuffled) {

            val chars = when(char) {
                'l' -> CharSets.lowercase
                'u' -> CharSets.uppercase
                'n' -> CharSets.numbers
                's' -> CharSets.symbols
                else -> characters
            }

            val random = Random.nextInt(0, chars.length - 1)
            password += chars[random]
        }

        return password
    }

    private fun updatePassword() {
        generatedPassword.value = generatePassword()
    }
}